File: files/elastic-solvers/ElasticityCommon.jl
"""Shared 2-D periodic elasticity routines. GPL-3.0-or-later."""
module ElasticityCommon

using FFTW, LinearAlgebra, Statistics

export PeriodicGrid, cubic, stress, reference_compliance, rms, prepare,
       diagnostics, package_result, plate_problem

function cubic(c11=4.0, c12=2.0, c44=1.0)
    minimum((c44, c11-c12, c11+2c12)) > 0 || error("Choose stable cubic constants.")
    stiffness = zeros(2,2,2,2)
    for i in 1:2, j in 1:2, k in 1:2, l in 1:2
        stiffness[i,j,k,l] = i == j == k == l ? c11 :
            (i == j && k == l ? c12 :
            ((i == k && j == l) || (i == l && j == k) ? c44 : 0.0))
    end
    stiffness
end

function stress(stiffness, strain)
    size(strain)[end-1:end] == (2,2) || error("Strain must end in two tensor axes.")
    result = zeros(size(strain))
    if ndims(stiffness) == 4
        for x in axes(strain,1), y in axes(strain,2), i in 1:2, j in 1:2, k in 1:2, l in 1:2
            result[x,y,i,j] += stiffness[i,j,k,l]*strain[x,y,k,l]
        end
    else
        for x in axes(strain,1), y in axes(strain,2), i in 1:2, j in 1:2, k in 1:2, l in 1:2
            result[x,y,i,j] += stiffness[x,y,i,j,k,l]*strain[x,y,k,l]
        end
    end
    result
end

function reference_compliance(reference, tensor)
    c11, c12, c44 = reference[1,1,1,1], reference[1,1,2,2], reference[1,2,1,2]
    result = similar(tensor)
    denominator = c11^2-c12^2
    result[:,:,1,1] = (c11*tensor[:,:,1,1]-c12*tensor[:,:,2,2])/denominator
    result[:,:,2,2] = (c11*tensor[:,:,2,2]-c12*tensor[:,:,1,1])/denominator
    result[:,:,1,2] = tensor[:,:,1,2]/(2c44)
    result[:,:,2,1] = result[:,:,1,2]
    result
end

struct PeriodicGrid
    size::Int
    wave_x::Matrix{Float64}
    wave_y::Matrix{Float64}
end

function PeriodicGrid(size=65)
    size >= 5 && isodd(size) || error("Use an odd grid size of at least 5.")
    frequencies = vcat(0:(size÷2), -(size÷2):-1) .* (2pi)
    PeriodicGrid(size, repeat(frequencies,1,size), repeat(frequencies',size,1))
end

transform(field) = fft(field,(1,2))
inverse(field) = real(ifft(field,(1,2)))

function strain(grid::PeriodicGrid, displacement)
    displacement_hat = transform(displacement)
    strain_hat = zeros(ComplexF64,grid.size,grid.size,2,2)
    wave = (grid.wave_x,grid.wave_y)
    for i in 1:2, j in 1:2
        strain_hat[:,:,i,j] = 0.5im .* (wave[j].*displacement_hat[:,:,i] + wave[i].*displacement_hat[:,:,j])
    end
    inverse(strain_hat)
end

function divergence(grid::PeriodicGrid, tensor)
    tensor_hat = transform(tensor)
    result_hat = zeros(ComplexF64,grid.size,grid.size,2)
    for i in 1:2
        result_hat[:,:,i] = im .* (grid.wave_x.*tensor_hat[:,:,i,1] + grid.wave_y.*tensor_hat[:,:,i,2])
    end
    inverse(result_hat)
end

function response(grid::PeriodicGrid, reference, polarization)
    polarization_hat = transform(polarization)
    displacement_hat = zeros(ComplexF64,grid.size,grid.size,2)
    for x in 1:grid.size, y in 1:grid.size
        k = (grid.wave_x[x,y],grid.wave_y[x,y])
        if x != 1 || y != 1
            acoustic = [sum(reference[i,j,m,l]*k[j]*k[l] for j in 1:2,l in 1:2) for i in 1:2,m in 1:2]
            force = [sum(polarization_hat[x,y,i,j]*k[j] for j in 1:2) for i in 1:2]
            displacement_hat[x,y,:] = -im .* (acoustic \ force)
        end
    end
    inverse(displacement_hat)
end

rms(array) = sqrt(mean(abs2,array))

function prepare(stiffness,eigenstrain,mean_strain,reference,relaxation,tolerance,max_iterations)
    size = axes(eigenstrain,1).stop
    grid = PeriodicGrid(size)
    mean_tensor = isnothing(mean_strain) ? zeros(2,2) : Matrix{Float64}(mean_strain)
    0 < relaxation <= 1 || error("Relaxation must lie in (0,1].")
    tolerance > 0 && max_iterations > 0 || error("Tolerance and iteration limit must be positive.")
    imposed = reshape(mean_tensor,1,1,2,2) .- eigenstrain
    initial_stress = stress(stiffness,imposed)
    scale = max(rms(divergence(grid,initial_stress)),2pi*rms(initial_stress),1e-30)
    grid, mean_tensor, scale
end

function diagnostics(grid,stiffness,eigenstrain,total_strain,scale)
    elastic_strain = total_strain-eigenstrain
    sigma = stress(stiffness,elastic_strain)
    energy_density = 0.5 .* dropdims(sum(elastic_strain.*sigma,dims=(3,4)),dims=(3,4))
    sigma,energy_density,rms(divergence(grid,sigma))/scale
end

function package_result(grid,stiffness,eigenstrain,total_strain,displacement,scale,history; extras...)
    sigma,energy_density,residual = diagnostics(grid,stiffness,eigenstrain,total_strain,scale)
    (; strain=total_strain,displacement,stress=sigma,energy_density,
       energy=mean(energy_density),residual,history=reduce(vcat,permutedims.(history)),extras...)
end

function plate_problem(;size=65,contrast=0.5,angle=30.0,misfit_ratio=-0.5,
                       matrix_constants=(4.0,2.0,1.0),precipitate_constants=nothing)
    PeriodicGrid(size)
    matrix = cubic(matrix_constants...)
    precipitate_values = isnothing(precipitate_constants) ? contrast .* collect(matrix_constants) : precipitate_constants
    precipitate = cubic(precipitate_values...)
    coordinate = collect(0:size-1)./size .- 0.5
    x = repeat(coordinate,1,size); y = repeat(coordinate',size,1)
    theta = deg2rad(angle)
    along = x.*cos(theta)+y.*sin(theta); across = -x.*sin(theta)+y.*cos(theta)
    phase = 0.5 .* (1 .- tanh.(((along./0.23).^2+(across./0.075).^2 .- 1)./0.35))
    stiffness = reshape(matrix,1,1,2,2,2,2) .+ reshape(phase,size,size,1,1,1,1).*reshape(precipitate-matrix,1,1,2,2,2,2)
    eigenstrain = zeros(size,size,2,2)
    eigenstrain[:,:,1,1] = 0.01phase; eigenstrain[:,:,2,2] = 0.01misfit_ratio.*phase
    stiffness,eigenstrain,0.5.*(matrix+precipitate),phase
end

end

File: files/elastic-solvers/elasticity_common.py
"""Transparent 2-D periodic elasticity helpers. GPL-3.0-or-later.
Arrays have shape (Nx, Ny, 2, 2). Both shear entries are stored:
epsilon_xy is tensor shear, so sigma_xy = 2*C44*epsilon_xy.
Use odd grids: they avoid even-grid Nyquist derivative ambiguities.
"""
import numpy as np


def cubic(c11=4.0, c12=2.0, c44=1.0):
    """In-plane cubic stiffness, with epsilon_zz^0 = epsilon_zz = 0.
    This is not a plane-stress reduction. Constants are dimensionless.
    """
    if not np.all(np.isfinite([c11, c12, c44])) or min(c44, c11-c12, c11+2*c12) <= 0:
        raise ValueError('Use stable cubic constants: C44>0, C11-C12>0, C11+2*C12>0.')
    stiffness = np.zeros((2, 2, 2, 2))
    for i in range(2):
        for j in range(2):
            for k in range(2):
                for l in range(2):
                    if i == j == k == l:
                        stiffness[i,j,k,l] = c11
                    elif i == j and k == l:
                        stiffness[i,j,k,l] = c12
                    elif (i == k and j == l) or (i == l and j == k):
                        stiffness[i,j,k,l] = c44
    return stiffness


def stress(stiffness, strain):
    """sigma_ij = sum_kl C_ijkl epsilon_kl; spatial axes are preserved."""
    return np.einsum('...ijkl,...kl->...ij', stiffness, strain)


def reference_compliance(reference, tensor):
    """Inverse on symmetric tensors, explicitly retaining tensor shear."""
    c11, c12, c44 = reference[0,0,0,0], reference[0,0,1,1], reference[0,1,0,1]
    result = np.empty_like(tensor)
    result[...,0,0] = (c11*tensor[...,0,0]-c12*tensor[...,1,1])/(c11*c11-c12*c12)
    result[...,1,1] = (c11*tensor[...,1,1]-c12*tensor[...,0,0])/(c11*c11-c12*c12)
    result[...,0,1] = tensor[...,0,1]/(2*c44)
    result[...,1,0] = result[...,0,1]
    return result


class PeriodicGrid:
    """Fourier differentiation on a unit square, x along array axis 0."""
    def __init__(self, size=65):
        if size < 5 or size % 2 != 1:
            raise ValueError('Choose an odd grid size of at least 5, e.g. 33, 65, 129.')
        self.size = size
        wave = 2*np.pi*np.fft.fftfreq(size, d=1/size)
        self.wave = np.stack(np.meshgrid(wave, wave, indexing='ij'), axis=-1)

    def transform(self, field):
        return np.fft.fftn(field, axes=(0,1))

    def inverse(self, field):
        return np.fft.ifftn(field, axes=(0,1)).real

    def strain(self, displacement):
        gradient_hat = 1j*np.einsum('...i,...j->...ij', self.transform(displacement), self.wave)
        return self.inverse((gradient_hat+gradient_hat.swapaxes(-1,-2))/2)

    def divergence(self, tensor):
        return self.inverse(1j*np.einsum('...ij,...j->...i', self.transform(tensor), self.wave))

    def response(self, reference, polarization):
        """Solve div(Cref:sym grad u - P)=0, mean(u)=0.
        Q_ik(k)=Cref_ijkl k_j k_l. u_hat=-i Q(k)^-1 P_hat.k.
        Full wavevectors are used: no missing |k|^-2 factor.
        """
        acoustic = np.einsum('ijkl,...j,...l->...ik', reference, self.wave, self.wave)
        acoustic[0,0] = np.eye(2)  # temporary only; displacement zero mode set below
        force_hat = np.einsum('...ij,...j->...i', self.transform(polarization), self.wave)
        displacement_hat = -1j*np.linalg.solve(acoustic, force_hat[...,None])[...,0]
        displacement_hat[0,0] = 0
        return self.inverse(displacement_hat)


def rms(array):
    return float(np.sqrt(np.mean(np.abs(array)**2)))


def prepare(stiffness, eigenstrain, mean_strain, reference, relaxation, tolerance, max_iterations):
    n = eigenstrain.shape[0]
    grid = PeriodicGrid(n)
    if eigenstrain.shape != (n,n,2,2) or stiffness.shape != (n,n,2,2,2,2):
        raise ValueError('Expected square fields of symmetric 2-D tensors.')
    if not 0 < relaxation <= 1 or tolerance <= 0 or max_iterations < 1:
        raise ValueError('Use 0 < relaxation <= 1, positive tolerance and iteration limit.')
    mean_strain = np.zeros((2,2)) if mean_strain is None else np.asarray(mean_strain,dtype=float)
    if mean_strain.shape != (2,2) or not np.allclose(mean_strain, mean_strain.T):
        raise ValueError('Mean strain must be a symmetric 2 by 2 tensor.')
    if not all(np.all(np.isfinite(a)) for a in [stiffness,eigenstrain,mean_strain,reference]):
        raise ValueError('All inputs must be finite.')
    # Normalize equilibrium by the initial forcing, retaining a finite scale
    # for uniform loading (whose initial divergence may be exactly zero).
    initial_stress = stress(stiffness, mean_strain-eigenstrain)
    scale = max(rms(grid.divergence(initial_stress)), 2*np.pi*rms(initial_stress), 1e-30)
    return grid, mean_strain, scale


def diagnostics(grid, stiffness, eigenstrain, strain, scale):
    elastic_strain = strain-eigenstrain
    sigma = stress(stiffness, elastic_strain)
    energy_density = 0.5*np.sum(elastic_strain*sigma, axis=(-2,-1))
    residual = rms(grid.divergence(sigma))/scale
    return sigma, energy_density, residual


def result(grid, stiffness, eigenstrain, strain, displacement, scale, history, **extra):
    sigma, energy_density, residual = diagnostics(grid,stiffness,eigenstrain,strain,scale)
    return dict(strain=strain, displacement=displacement, stress=sigma,
                energy_density=energy_density, energy=float(energy_density.mean()),
                residual=residual, history=np.asarray(history), **extra)


def plate_problem(size=65, contrast=0.5, angle=30.0, misfit_ratio=-0.5,
                  matrix_constants=(4.,2.,1.), precipitate_constants=None):
    """A smooth periodic plate. angle is its long-axis angle from x, in degrees.
    Rotate the shape, keeping stiffness and physical eigenstrain crystal-fixed.
    Optional independent precipitate_constants override scalar contrast.
    """
    PeriodicGrid(size)
    matrix = cubic(*matrix_constants)
    precipitate = cubic(*(np.asarray(matrix_constants)*contrast if precipitate_constants is None else precipitate_constants))
    x,y = np.meshgrid(np.arange(size)/size-.5,np.arange(size)/size-.5,indexing='ij')
    theta = np.deg2rad(angle)
    along = x*np.cos(theta)+y*np.sin(theta)
    across = -x*np.sin(theta)+y*np.cos(theta)
    radius_squared = (along/.23)**2+(across/.075)**2
    # Smooth fixed interface width in physical coordinates, not in pixels.
    phase = .5*(1-np.tanh((radius_squared-1)/.35))
    stiffness = matrix+phase[...,None,None,None,None]*(precipitate-matrix)
    eigenstrain = phase[...,None,None]*np.diag([.01,.01*misfit_ratio])
    reference = .5*(matrix+precipitate)
    return stiffness,eigenstrain,reference,phase


def plot_solution(solution, phase):
    """Plot geometry, physical fields and residual, with readable rainbow maps."""
    import matplotlib.pyplot as plt
    figure, axes = plt.subplots(1,4,figsize=(15,3.5),layout='constrained')
    for axis,field,title in zip(axes[:3], [phase,solution['stress'][...,0,0],solution['energy_density']],
                               ['Precipitate fraction h',r'$\sigma_{xx}$','Elastic energy density']):
        image=axis.imshow(field.T,origin='lower',extent=[0,1,0,1],cmap='rainbow')
        axis.set(title=title,xlabel='x',ylabel='y')
        figure.colorbar(image,ax=axis,shrink=.8)
    history=solution['history']
    axes[3].semilogy(history[:,0],np.maximum(history[:,1],1e-16))
    axes[3].set(xlabel='Iteration',ylabel='Equilibrium residual',title='Convergence')
    return figure
